Привести в соответствие с ТЗ спринта № 3

Phronima6 opened the pull request:
Привести в соответствие с ТЗ спринта № 3

Phronima6 committed:
Привести в соответствие с ТЗ спринта № 3

Phronima6 committed:
Устранить замечания для корректной работы программы

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/comment/repository/CommentRepository.java:
```diff
@@ -0,0 +1,13 @@
+import ru.yandex.practicum.comment.model.Comment;
+import java.util.List;
+
+public interface CommentRepository extends JpaRepository<Comment, Long> {
```
По требованиям Д/З нужно использовать Spring Data JDBC

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/image/repository/ImageRepository.java:
```diff
@@ -0,0 +1,7 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import ru.yandex.practicum.image.model.Image;
+
+public interface ImageRepository extends JpaRepository<Image, Long> {
```
По требованиям Д/З нужно использовать Spring Data JDBC

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/like/repository/LikeRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import ru.yandex.practicum.like.model.Like;
+
+public interface LikeRepository extends JpaRepository<Like, Long> {
```
По требованиям Д/З нужно использовать Spring Data JDBC

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/post/service/PostServiceImplements.java:
```diff
@@ -0,0 +1,176 @@
+    public List<PostDtoResponseShort> getPosts() {
+        log.info("Попытка найти все посты.");
+        return postRepository.findAll().stream()
+                .map((post -> {
+                    final PostDtoResponseShort postDtoResponseShort = postMapper.toPostDtoResponseShort(post);
+                    final Image image = imageService.getImage(post.getImageId());
+                    final String imageBase64 = Base64.getEncoder().encodeToString(image.getImageData());
+                    postDtoResponseShort.setBase64Image(imageBase64);
+                    postDtoResponseShort.setPostPreview(getPreview(post.getPostText()));
+                    postDtoResponseShort.setCountLikes(likeService.getCountLikesOfPost(post.getPostId()));
+                    postDtoResponseShort.setTagsDto(tagService.findAllTagsByPostId(post.getPostId()));
+                    postDtoResponseShort.setCountComments(commentService.getCountCommentsOfPost(post.getPostId()));
+                    return postDtoResponseShort;
```
В индустрии есть сложившаяся практика держать лямбды не более 4-5 строк для читабельности. Эту лямбду лучше будет вынести в отдельный приватный метод

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/post/service/PostServiceImplements.java:
```diff
@@ -0,0 +1,176 @@
+        final List<Long> postIds = tagService.findAllPostIdByTagText(tagText);
+        log.info("По тегу найдены посты со следующими id: {}.", postIds);
+        return postRepository.findByPostIdIn(postIds).stream()
+                .map((post -> {
+                    final PostDtoResponseShort postDtoResponseShort = postMapper.toPostDtoResponseShort(post);
+                    final Image image = imageService.getImage(post.getImageId());
+                    final String imageBase64 = Base64.getEncoder().encodeToString(image.getImageData());
+                    postDtoResponseShort.setBase64Image(imageBase64);
+                    postDtoResponseShort.setPostPreview(getPreview(post.getPostText()));
+                    postDtoResponseShort.setCountLikes(likeService.getCountLikesOfPost(post.getPostId()));
+                    postDtoResponseShort.setTagsDto(tagService.findAllTagsByPostId(post.getPostId()));
+                    postDtoResponseShort.setCountComments(commentService.getCountCommentsOfPost(post.getPostId()));
+                    return postDtoResponseShort;
+                }))
```
Дублирование кода

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/post/service/PostServiceImplements.java:
```diff
@@ -0,0 +1,176 @@
+        final Pageable pageable = PageRequest.of(page - 1, size, Sort.by("postId").descending());
+        final Page<Post> postPage = postRepository.findAll(pageable);
+        return postPage.map(post -> {
+            final PostDtoResponseShort postDtoResponseShort = postMapper.toPostDtoResponseShort(post);
+            final Image image = imageService.getImage(post.getImageId());
+            final String imageBase64 = Base64.getEncoder().encodeToString(image.getImageData());
+            postDtoResponseShort.setBase64Image(imageBase64);
+            postDtoResponseShort.setPostPreview(getPreview(post.getPostText()));
+            postDtoResponseShort.setCountLikes(likeService.getCountLikesOfPost(post.getPostId()));
+            postDtoResponseShort.setTagsDto(tagService.findAllTagsByPostId(post.getPostId()));
+            postDtoResponseShort.setCountComments(commentService.getCountCommentsOfPost(post.getPostId()));
+            return postDtoResponseShort;
```
Дублирование кода

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/post/service/PostServiceImplements.java:
```diff
@@ -0,0 +1,176 @@
+                .orElseThrow(() -> new NotFoundException("Пост с id: " + postId + " не найден."));
+    }
+
+    @Transactional(readOnly = true)
+    private static String getPreview(final String postText) {
```
Spring использует прокси для добавления логики, а они работают через наследование/имплементацию и переопределение методов. Статические и приватные методы переопределять, увы, нельзя

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/post/repository/PostRepository.java:
```diff
@@ -0,0 +1,11 @@
+import ru.yandex.practicum.post.model.Post;
+import java.util.List;
+
+public interface PostRepository extends JpaRepository<Post, Long> {
```
По требованиям Д/З нужно использовать Spring Data JDBC

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/tag/repository/TagRepository.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.List;
+import java.util.Set;
+
+public interface TagRepository extends JpaRepository<Tag, Long> {
```
По требованиям Д/З нужно использовать Spring Data JDBC

Phronima6 replied:
Привет! Спасибо, все поправлю!

Phronima6 committed:
Привести в соответствие с замечаниями:
- использовать Spring Data JDBC вместо JpaRepository;
- вынести метод getPreview (класса PostServiceImplements) в Interface;
- объединить дублирующийся код в классе PostServiceImplements в отдельный приватный метод (создан с названием buildPostDtoResponseShort)

yndx-artur-enikeev commented on src/main/java/ru/yandex/practicum/comment/controller/CommentController.java:
```diff
@@ -0,0 +1,46 @@
+
+    CommentService commentService;
+
+    @Autowired
+    public CommentController(final CommentService commentService) {
+        this.commentService = commentService;
+    }
```
Не замечание, просто комментарий: если конструктор один, то `@Autowired` указывать необязательно, Spring поймет. И если в конструкторе нет вызова `super()` или какой-либо логики помимо инициализации полей, то можно использовать `@RequiredArgsConstructor` или `@AllArgsConstructor`, чтобы Lombok сгенерировал конструктор

yndx-artur-enikeev commented on src/test/java/practicum/comment/controller/CommentControllerTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
```
openMocks возвращает AutoClosable объект, Javadocs к этому методу рекомендуют этот объект сохранять в поле и вызывать у него метод close() в `@AfterEach`

Phronima6 replied:
Спасибо! Исправлю в следующей версии.
